### PR TITLE
Fixing FlashAttention ModernBert.

### PR DESCRIPTION
**Pull request by Narsil**
# What does this PR do?

Fixes ModernBert to use flash attention when possible.

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by Narsil** [comment]
@kozistr I've implemented the flash attention version.

I tried to keep things at a minimum here, but I think we can still vastly improve the various caching.

File: backends/candle/src/models/flash_modernbert.rs [hunk 1]
```diff
@@ -1,14 +1,15 @@
 use std::collections::HashMap;
 
 use crate::flash_attn::flash_attn_varlen;
-use crate::layers::{apply_rotary, get_cos_sin, get_inv_freqs, LayerNorm, Linear};
+use crate::layers::{get_cos_sin, get_inv_freqs, LayerNormNoBias, Linear};
```
Comment (Narsil) [inline backends/candle/src/models/flash_modernbert.rs:4]: This was already landed in the orginal PR, it's a good way to handle the fact that the layer norm never has any bias here.

File: backends/candle/src/models/flash_modernbert.rs [hunk 2]
```diff
@@ -79,35 +80,34 @@ impl ModernBertAttention {
         new_qkv_shape.pop();
         new_qkv_shape.push(self.num_attention_heads * 3);
         new_qkv_shape.push(self.attention_head_size);
-        let qkv = qkv.reshape(new_qkv_shape.as_slice())?.transpose(1, 2)?;
+        let qkv = qkv.reshape(new_qkv_shape.as_slice())?;
```
Comment (Narsil) [inline backends/candle/src/models/flash_modernbert.rs:83]: No transpose, no contiguous, taken from other flash implementations.

File: backends/candle/src/models/flash_modernbert.rs [hunk 3]
```diff
@@ -79,35 +80,34 @@ impl ModernBertAttention {
...
+        let q = qkv.narrow(1, 0, self.num_attention_heads)?;
+        let k = qkv.narrow(1, self.num_attention_heads, self.num_attention_heads)?;
+        let v = qkv.narrow(1, self.num_attention_heads * 2, self.num_attention_heads)?;
 
-        let query_layer = apply_rotary(query_layer, cos, sin, self.attention_head_size)?;
-        let key_layer = apply_rotary(key_layer, cos, sin, self.attention_head_size)?;
+        apply_rotary_inplace(&q, &k, &cos, &sin, true)?;
 
-        let attention_size = if self.use_local_attention {
+        let window_size = if self.use_local_attention {
```
Comment (Narsil) [inline backends/candle/src/models/flash_modernbert.rs:92]: Rename for clarity, in flash attention, this is called window_size.

File: backends/candle/src/models/flash_modernbert.rs [hunk 4]
```diff
@@ -277,13 +276,22 @@ impl FlashModernBertModel {
...
         };
 
-        let embeddings = ModernBertEmbeddings::load(vb.pp("model.embeddings"), config)?;
-        let encoder = ModernBertEncoder::load(vb.pp("model.layers"), config)?;
-        let final_norm = LayerNorm::load(
+        let embeddings = ModernBertEmbeddings::load(vb.pp("model.embeddings"), config)
+            .or_else(|_| ModernBertEmbeddings::load(vb.pp("embeddings"), config))?;
+        let encoder = ModernBertEncoder::load(vb.pp("model.layers"), config)
+            .or_else(|_| ModernBertEncoder::load(vb.pp("layers"), config))?;
+        let final_norm = LayerNormNoBias::load(
```
Comment (Narsil) [inline backends/candle/src/models/flash_modernbert.rs:283]: Handle various names like the non flash modeling.

File: backends/candle/src/models/flash_modernbert.rs [hunk 5]
```diff
@@ -295,15 +303,11 @@ impl FlashModernBertModel {
                 config.global_rope_theta
             };
 
-            let max_position_embeddings = if use_local_attention {
-                config.max_position_embeddings
-            } else {
-                config.local_attention
-            };
+            let max_position_embeddings = config.max_position_embeddings;
```
Comment (Narsil) [inline backends/candle/src/models/flash_modernbert.rs:306]: This should model what was there before.

File: backends/candle/src/models/flash_modernbert.rs [hunk 6]
```diff
@@ -295,15 +303,11 @@ impl FlashModernBertModel {
...
 
             let inv_freqs = get_inv_freqs(rotary_dim, rope_theta as f32, vb.device(), None)?;
 
-            let (cos, sin) = get_cos_sin(max_position_embeddings, &inv_freqs, vb.dtype(), true)?;
+            let (cos, sin) = get_cos_sin(max_position_embeddings, &inv_freqs, vb.dtype(), false)?;
```
Comment (Narsil) [inline backends/candle/src/models/flash_modernbert.rs:310]: Very important, in flash models this is false, it's true in non flash models.

Took me quite a bit to see this difference :(

File: backends/candle/src/models/flash_modernbert.rs [hunk 7]
```diff
@@ -343,9 +346,6 @@ impl FlashModernBertModel {
             let cos = cos.index_select(&position_ids, 0)?;
             let sin = sin.index_select(&position_ids, 0)?;
```
Comment (Narsil) [inline backends/candle/src/models/flash_modernbert.rs:348]: Not necessary with the fast rotary kernel.

File: backends/candle/src/models/modernbert.rs [hunk 8]
```diff
@@ -578,7 +578,10 @@ impl ModernBertModel {
     }
 
     fn get_local_attention_mask(&self, attention_mask: &Tensor) -> Result<Tensor> {
-        let attention_mask = attention_mask.to_dtype(DType::U8)?;
+        let dev = attention_mask.device();
+        let attention_mask = attention_mask
```
Comment (Narsil) [inline backends/candle/src/models/modernbert.rs:582]: In order for the non flash code to work on GPU (when flash isn't available for instance).

Then we cannot do the `.abs()` operation (it's not in candle). Doing the mask on CPU is acceptable for now (GPU + non flash should be slightly rare nowadays).

File: backends/grpc-client/build.rs [hunk 9]
```diff
@@ -1,7 +1,7 @@
 use std::fs;
 
 fn main() -> Result<(), Box<dyn std::error::Error>> {
-    println!("cargo:rerun-if-changed=../../proto/embed.proto");
+    println!("cargo:rerun-if-changed=../proto/embed.proto");
```
Comment (Narsil) [inline backends/grpc-client/build.rs:4]: This causes to many rebuilds because the file location is incorrect.

File: backends/candle/src/lib.rs [hunk 10]
```diff
@@ -357,6 +358,25 @@ impl CandleBackend {
                 }
             }
             #[cfg(feature = "cuda")]
+            (Config::ModernBert(config), Device::Cuda(_)) => {
+                if cfg!(any(feature = "flash-attn", feature = "flash-attn-v1"))
+                    && dtype == DType::F16
+                    // Allow disabling because of flash attention v1 precision problems
+                    // See: https://github.com/huggingface/text-embeddings-inference/issues/37
+                    && &std::env::var("USE_FLASH_ATTENTION").unwrap_or("True".to_string()).to_lowercase() == "true"
```
Comment (ivarflakstad) [inline backends/candle/src/lib.rs:366]: Do we want to communicate to the user why they are not running with  FA enabled?

**Review by ivarflakstad — APPROVED** [review]
I haven't tested these changes myself, but everything LGTM 👌

File: backends/candle/src/lib.rs (hunk 10 above)
Comment (Narsil) [reply backends/candle/src/lib.rs:366]: I took this code from the previous model. The logs *do* convey if it's flashModernBert or Modernbert.

Comment (kozistr) [reply backends/candle/src/lib.rs:366]: It's a nit. How about adding an error message or comment here additionally that `FlashModernBert` does not support `flash-attn-v1` due to [the lack of `attention windowing` feature](https://github.com/huggingface/text-embeddings-inference/blob/510423698d902b849ec89ee85382582cd4a6df65/backends/candle/src/flash_attn.rs#L43)?

**Review by kozistr — COMMENTED** [review]
Thanks for your work! I leave some minor comments, and others look great to me :)

File: backends/candle/src/lib.rs (hunk 10 above)
Comment (Narsil) [reply backends/candle/src/lib.rs:366]: Great comment, I'll change that around.

**Review by kozistr — APPROVED** [review]
LGTM!
